include/alpaka/queue/cuda_hip/QueueUniformCudaHipRt.hpp

```diff
@@ -0,0 +1,298 @@
+        //!
+        //! Blocks execution of the calling thread until the queue has finished processing all previously requested
+        //! tasks (kernels, data copies, ...)
+        template<bool TBlocking>
+        struct CurrentThreadWaitFor<uniform_cuda_hip::detail::QueueUniformCudaHipRt<TBlocking>>
+        {
+            ALPAKA_FN_HOST static auto currentThreadWaitFor(
+                uniform_cuda_hip::detail::QueueUniformCudaHipRt<TBlocking> const& queue) -> void
+            {
+                ALPAKA_DEBUG_MINIMAL_LOG_SCOPE;
+
+                // Sync is allowed even for queues on non current device.
+                ALPAKA_UNIFORM_CUDA_HIP_RT_CHECK(ALPAKA_API_PREFIX(StreamSynchronize)(queue.getNativeHandle()));
+            }
+        };
+
+        //! The CUDA/HIP RT queue enqueue trait specialization.
+        template<bool TBlocking, typename TTask>
+        struct Enqueue<uniform_cuda_hip::detail::QueueUniformCudaHipRt<TBlocking>, TTask>
+        {
+            enum class CallbackState
+            {
+                enqueued,
+                notified,
+                finished,
+            };
+
+            struct CallbackSynchronizationData : public std::enable_shared_from_this<CallbackSynchronizationData>
+            {
+                std::mutex m_mutex;
+                std::condition_variable m_event;
+                CallbackState m_state = CallbackState::enqueued;
+            };
+
+#    if defined(ALPAKA_ACC_GPU_CUDA_ENABLED)
+            static void CUDART_CB
+#    else
+            static void HIPRT_CB
+#    endif
+            uniformCudaHipRtCallback(
+                ALPAKA_API_PREFIX(Stream_t) /*queue*/,
+                ALPAKA_API_PREFIX(Error_t) /*status*/,
+                void* arg)
+            {
+                // explicitly copy the shared_ptr so that this method holds the state even when the executing thread
+                // has already finished.
+                const auto pCallbackSynchronizationData
+                    = reinterpret_cast<CallbackSynchronizationData*>(arg)->shared_from_this();
+
+                // Notify the executing thread.
+                {
+                    std::unique_lock<std::mutex> lock(pCallbackSynchronizationData->m_mutex);
+                    pCallbackSynchronizationData->m_state = CallbackState::notified;
+                }
+                pCallbackSynchronizationData->m_event.notify_one();
+
+                // Wait for the executing thread to finish the task if it has not already finished.
+                std::unique_lock<std::mutex> lock(pCallbackSynchronizationData->m_mutex);
+                if(pCallbackSynchronizationData->m_state != CallbackState::finished)
+                {
+                    pCallbackSynchronizationData->m_event.wait(
+                        lock,
+                        [pCallbackSynchronizationData]()
+                        { return pCallbackSynchronizationData->m_state == CallbackState::finished; });
```

Comment:
```suggestion
                        [pCallbackSynchronizationData]
                        { return pCallbackSynchronizationData->m_state == CallbackState::finished; });
```

Reply:
No, I am not removing the parentheses.

Reply:
I didn't know, that is allowed. I think, it is not well known, that is allowed. Therefore I would also keep the parentheses.

Reply:
Yes, it is allowed, but IMHO it makes the (already confusing) lambda syntax even more confusing.

Reply:
> No, I am not removing the parentheses.

I don't insist on it. I would be interested in your reasoning though. Why keep them?

Reply:
As I wrote above, IMHO the lambda syntax is more confusing without them.

Reply:
Because, if you are not familiar with the syntax, how do find out, what the syntax mean? If you didn't had the idea of a lambda, try to ask a search engine, what `[]{}` means in c++.

I have similar problems in Python.

Reply:
Ok, you convinced me. Keep the parentheses.